### PR TITLE
[CLDC-2068] Add UPRN questions to lettings

### PR DESCRIPTION
**Pull request by ghost**
Add UPRN questions to Letting Logs and add question numbers to both sales and lettings.

https://digital.dclg.gov.uk/jira/browse/CLDC-2068

## For testing:

valid UPRN: 100023336963
invalid UPRN: 100023336961

### CODE REVIEW

File: app/models/derived_variables/lettings_log_variables.rb [hunk 1]
```diff
@@ -94,6 +94,17 @@ def set_derived_fields!
...
+    if uprn_known&.zero?
+      self.uprn = nil
+    end
+
+    if uprn_confirmed&.zero?
+      self.uprn = nil
+      self.uprn_known = 0
+    end
+
+    reset_address_fields! if needstype == 2
```
Comment (jamesrose) [inline app/models/derived_variables/lettings_log_variables.rb]:
```suggestion
    reset_address_fields! if is_supported_housing?
```

File: app/models/form/lettings/pages/address.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      Form::Lettings::Questions::AddressLine2.new(nil, nil, self),
+      Form::Lettings::Questions::TownOrCity.new(nil, nil, self),
+      Form::Lettings::Questions::County.new(nil, nil, self),
+      Form::Lettings::Questions::PostcodeForFullAddress.new(nil, nil, self),
+    ]
+  end
+
+  def routed_to?(log, _current_user = nil)
+    return false if log.uprn_known.nil?
+    return false if log.needstype == 2
```
Comment (jamesrose) [inline app/models/form/lettings/pages/address.rb]:
```suggestion
    return false if log.is_supported_housing?
```

File: app/models/form/lettings/pages/property_local_authority.rb [hunk 3]
```diff
@@ -8,4 +8,12 @@ def initialize(id, hsh, subsection)
   def questions
     @questions ||= [Form::Lettings::Questions::La.new(nil, nil, self)]
   end
+
+  def routed_to?(log, _current_user = nil)
+    return false if log.uprn_known.nil? && form.start_date.year >= 2023
+    return false if log.is_la_inferred?
+    return false if log.needstype == 2
```
Comment (jamesrose) [inline app/models/form/lettings/pages/property_local_authority.rb]:
```suggestion
    return false if log.is_supported_housing?
```

File: app/models/form/lettings/pages/uprn.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def questions
+    @questions ||= [
+      Form::Lettings::Questions::Uprn.new(nil, nil, self),
+    ]
+  end
+
+  def routed_to?(log, _current_user = nil)
+    return false if log.needstype == 2
```
Comment (jamesrose) [inline app/models/form/lettings/pages/uprn.rb]:
```suggestion
    return false if log.is_supported_housing?
```

File: app/models/form/lettings/pages/uprn_known.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  end
+
+  def questions
+    @questions ||= [
+      Form::Lettings::Questions::UprnKnown.new(nil, nil, self),
+    ]
+  end
+
+  def routed_to?(log, _current_user = nil)
+    log.needstype != 2
```
Comment (jamesrose) [inline app/models/form/lettings/pages/uprn_known.rb]:
```suggestion
    !log.is_supported_housing?
```

Comment (kosiakkatrina) [reply app/models/form/lettings/pages/uprn_known.rb]: Can it be `general_needs?` instead? (I don't know if this method exists btw)

File: app/models/form/lettings/questions/address_line1.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    @id = "address_line1"
+    @check_answer_label = "Address"
+    @header = "Address line 1"
+    @type = "text"
+    @plain_label = true
+    @question_number = 12
+  end
+
+  def hidden_in_check_answers?(log, _current_user = nil)
+    return true if log.uprn_known.nil?
```
Comment (kosiakkatrina) [inline app/models/form/lettings/questions/address_line1.rb:13]: Do we need this? Would the page be routed to if uprn_known is not answered?

Comment (ghost) [reply app/models/form/lettings/questions/address_line1.rb:13]: probably redundant!

File: app/models/form/lettings/pages/uprn_known.rb (hunk 5 above)
Comment (ghost) [reply app/models/form/lettings/pages/uprn_known.rb]: they're not equivalent; one evaluates `true` when `needstype == nil` and the other doesn't

**Review by jamesrose — APPROVED** [review]

**Pull request merged**